rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -1,226 +1,370 @@
+    from rasa.shared.nlu.training_data.features import Features
+
+    if bilou_tagging:
+        _tags = bilou_tags_to_ids(example, tag_spec.tags_to_ids, tag_spec.tag_name)
+    else:
+        _tags = []
+        for token in example.get(TOKENS_NAMES[TEXT]):
+            _tag = determine_token_labels(
+                token, example.get(ENTITIES), attribute_key=tag_spec.tag_name
+            )
+            _tags.append(tag_spec.tags_to_ids[_tag])
 
-MASK = "mask"
+    # transpose to have seq_len x 1
+    return Features(
+        np.array([_tags]).T, FEATURE_TYPE_SEQUENCE, tag_spec.tag_name, TAG_ID_ORIGIN
+    )
 
 
-def surface_attributes(
-    tracker_state_features: List[List[Dict[Text, List["Features"]]]]
+def _surface_attributes(
+    features: List[List[Dict[Text, List["Features"]]]],
+    featurizers: Optional[List[Text]] = None,
 ) -> Dict[Text, List[List[List["Features"]]]]:
     """Restructure the input.
 
+    "features_for_examples" can, for example, be a dictionary of attributes (INTENT,
+    TEXT, ACTION_NAME, ACTION_TEXT, ENTITIES, SLOTS, FORM) to a list of features for
+    all dialogue turns in all training trackers.
+    For NLU training it would just be a dictionary of attributes (either INTENT or
+    RESPONSE, TEXT, and potentially ENTITIES) to a list of features for all training
+    examples.
+
     Args:
-        tracker_state_features: a dictionary of attributes (INTENT, TEXT, ACTION_NAME,
-            ACTION_TEXT, ENTITIES, SLOTS, FORM) to a list of features for all
-            dialogue turns in all training trackers
+        features: a dictionary of attributes to a list of features for all
+            examples in the training data
+        featurizers: the featurizers to consider
 
     Returns:
-        A dictionary of attributes to a list of features for all dialogue turns
-        and all training trackers.
+        A dictionary of attributes to a list of features for all examples.
     """
     # collect all attributes
     attributes = set(
         attribute
-        for features_in_tracker in tracker_state_features
-        for features_in_turn in features_in_tracker
-        for attribute in features_in_turn.keys()
+        for list_of_attribute_to_features in features
+        for attribute_to_features in list_of_attribute_to_features
+        for attribute in attribute_to_features.keys()
     )
 
-    attribute_to_features = defaultdict(list)
+    output = defaultdict(list)
 
-    for features_in_tracker in tracker_state_features:
+    for list_of_attribute_to_features in features:
         intermediate_features = defaultdict(list)
-
-        for features_in_dialogue in features_in_tracker:
+        for attribute_to_features in list_of_attribute_to_features:
             for attribute in attributes:
+                features = attribute_to_features.get(attribute)
+                if featurizers:
+                    featurizers = _filter_features(features, featurizers)

```

Comment:
Shouldn't this be `features`, since you are manipulating the list of features inside `_filter_features` and returning it.

Reply:
@tabergma could you please take this over